WordPress/gutenberg
# Enhance the shortcode block to support previewing of shortcodes

## Description

This PR adds the capability to preview shortcodes entered in the Shortcode block. This is similar to previews rendered by the embed block and custom HTML block. Work done in this PR references the issue #1054 

## Summary of changes

- I've  added a new REST API endpoint that converts the shortcode to its output filtered through its hooks
    - `GET /gutenberg/v1/shortcodes` - get the filtered output for a given shortcode

- I've changed the UI of the shortcode block to a tabbed component (identical to the custom HTML block).  I've posted a [comment](https://github.com/WordPress/gutenberg/issues/1054#issuecomment-360271233) on my thoughts behind this design consideration. The comment has screenshots of the UI, please check there. 

- The shortcode preview is embedded in a `Sandbox` component. This component (which is essentially creating an iframe) needed to support injection of custom JS and styles, since some shortcodes such as [gallery] and [playlist] need  additional styles and scripts, such as the theme style or mediaelement JS. I've made the necessary changes to support this.

## How to test

1. Test with all the default shortcodes - `[audio]`, `[video]`, `[caption]`, `[wp_caption]`, `[embed]`, `[gallery]`, `[playlist type="audio"]`, `[playlist type="video"]`.
2. Add various attributes to the shortcodes and check if they get applied. To note, shortcodes entered without any src attributes will try to automatically render an attached media  file of the corresponding type. For eg. : 
           (a) [audio] will render the attached audio vs [audio src="audio-source.mp3"]
           (b) [gallery] will render all attached images, whereas [gallery ids="10,11,12"] will render those specific image ids.
3. Create a custom shortcode via add_shortcode and use it in the shortcode block. Also, get a little fancy with [Shortcode-ception](https://developer.wordpress.org/plugins/shortcodes/enclosing-shortcodes/#shortcode-ception).
4. Edit/modify/replace shortcodes in the block - clicking preview after every such action should render the most recent version of the shortcode
5. Non-existent shortcodes will be treated as a normal text and displayed
6. If there's no attached media/photo to a post, then using just [gallery] or [audio] (without "src" or "ids" attributes) will give an error message that the preview cannot be rendered
7. Giving incompatible "ids" - for example, in [gallery ids="11"], if 11 refers to an audio file, or in [playlist ids="12" type="video"], "12" refers to an audio file, then a message will be flashed "Sorry, couldn't render preview" 

## Screenshot

Below is an example of a preview for [caption]

![giphy](https://user-images.githubusercontent.com/1269602/35482917-96a141a0-0461-11e8-9afb-af58e584ce0b.gif)

## <del>Known issue</del>

<del>1. Preview for [wp-playlist] does not work. I'm trying to inject all the necessary JS into the Sandbox component (the js is fetched by the rest endpoint and can be seen in these [lines of code](https://github.com/niranjan-uma-shankar/gutenberg/blob/master/lib/class-wp-rest-shortcodes-controller.php#L110)). I get an error `Uncaught TypeError: Cannot read property 'replace' of undefined` in `underscore.min.js`. Any pointers here would be appreciated.</del>
